### PR TITLE
[mlir][llvm] Return failure from type converter for n-D scalable vectors

### PR DESCRIPTION
**Pull request by c-rhodes**
This patch changes vector type conversion to return failure on n-D scalable vector types instead of asserting.

This is an alternative approach to #65261 that aims to enable lowering of Vector ops directly to ArmSME intrinsics where possible, and seems more consistent with other type conversions. It's trivial to hit the assert at the moment and it could be interpreted as n-D scalable vector types being a bug, when they're valid types in the Vector dialect.

By returning failure it will generally fail more gracefully, particularly for release builds or other builds where assertions are disabled.

### CODE REVIEW

**Comment by banach-space** [comment]
Thank you, this is far less intrusive and feels much more canonical than https://github.com/llvm/llvm-project/pull/65261 - this would be my preference :)

This should be sufficient for two types of lowerings from `Vector` to `ArmSME`:
1. Vector -> ArmSME Ops -> LLVM SME intrinsics (_with_ custom ops),
2. Vector -> LLVM SME intrinsics (_without_ custom ops).

ATM, we only use 1. I am guessing that you'd like to use 2. for `vector.outerproduct` to lower to SME's MOPA instructions, e.g. [FMOPA](https://developer.arm.com/documentation/ddi0602/2021-06/SME-Instructions/FMOPA--non-widening---Floating-point-outer-product-and-accumulate-)? I am still wondering whether mixing 1. and 2. is the right approach.

But that's something we can discuss in a different patch,, this change makes sense regardless.

LGTM, but please wait for more reviews before landing this :)

**Review by banach-space — APPROVED** [review]

**Comment by c-rhodes** [comment]
> ATM, we only use 1. I am guessing that you'd like to use 2. for `vector.outerproduct` to lower to SME's MOPA instructions, e.g. [FMOPA](https://developer.arm.com/documentation/ddi0602/2021-06/SME-Instructions/FMOPA--non-widening---Floating-point-outer-product-and-accumulate-)? I am still wondering whether mixing 1. and 2. is the right approach.
> 
> But that's something we can discuss in a different patch,, this change makes sense regardless.

That's correct, I want to lower `vector.outerproduct` directly to intrinsics, at least initially as this seems to work. But this would also open this up for other ops where possible.

> 
> LGTM, but please wait for more reviews before landing this :)

Thanks for reviewing!

File: mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp
Comment (dcaballe) [inline mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp:510]: Do we handle the 0-D case gracefully here?

**Review by dcaballe — APPROVED** [review]
LGTM, thanks! Yeah, much better!

File: mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp
Comment (c-rhodes) [reply mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp:510]: For 0-D it won't reach here having returned already

**Comment by c-rhodes** [comment]
Thanks for reviewing!

**Pull request merged**
